// folio/out/config.js
"use strict";
/**
 * Copyright Microsoft Corporation. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultConfig = void 0;
const path_1 = __importDefault(require("path"));
exports.defaultConfig = {
    forbidOnly: false,
    globalTimeout: 0,
    grep: '.*',
    maxFailures: 0,
    outputDir: path_1.default.join(process.cwd(), 'test-results'),
    quiet: false,
    repeatEach: 1,
    retries: 0,
    shard: undefined,
    snapshotDir: '__snapshots__',
    testDir: '',
    timeout: 10000,
    updateSnapshots: false,
    workers: Math.ceil(require('os').cpus().length / 2),
};
