// folio/out/transform.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.installTransform = void 0;
const crypto = __importStar(require("crypto"));
const os = __importStar(require("os"));
const path = __importStar(require("path"));
const fs = __importStar(require("fs"));
const pirates = __importStar(require("pirates"));
const babel = __importStar(require("@babel/core"));
const sourceMapSupport = __importStar(require("source-map-support"));
const version = 3;
const cacheDir = path.join(os.tmpdir(), 'playwright-transform-cache');
const sourceMaps = new Map();
sourceMapSupport.install({
    environment: 'node',
    handleUncaughtExceptions: false,
    retrieveSourceMap(source) {
        if (!sourceMaps.has(source))
            return null;
        const sourceMapPath = sourceMaps.get(source);
        if (!fs.existsSync(sourceMapPath))
            return null;
        return {
            map: JSON.parse(fs.readFileSync(sourceMapPath, 'utf-8')),
            url: source
        };
    }
});
function calculateCachePath(content, filePath) {
    const hash = crypto.createHash('sha1').update(content).update(filePath).update(String(version)).digest('hex');
    const fileName = path.basename(filePath, path.extname(filePath)).replace(/\W/g, '') + '_' + hash;
    return path.join(cacheDir, hash[0] + hash[1], fileName);
}
function installTransform() {
    return pirates.addHook((code, filename) => {
        const cachePath = calculateCachePath(code, filename);
        const codePath = cachePath + '.js';
        const sourceMapPath = cachePath + '.map';
        sourceMaps.set(filename, sourceMapPath);
        if (fs.existsSync(codePath))
            return fs.readFileSync(codePath, 'utf8');
        const result = babel.transformFileSync(filename, {
            presets: [
                ['@babel/preset-env', { targets: { node: '10.17.0' } }],
                ['@babel/preset-typescript', { onlyRemoveTypeImports: true }],
            ],
            plugins: [['@babel/plugin-proposal-class-properties', { loose: true }]],
            sourceMaps: 'both',
        });
        if (result.code) {
            fs.mkdirSync(path.dirname(cachePath), { recursive: true });
            if (result.map)
                fs.writeFileSync(sourceMapPath, JSON.stringify(result.map), 'utf8');
            fs.writeFileSync(codePath, result.code, 'utf8');
        }
        return result.code;
    }, {
        exts: ['.ts']
    });
}
exports.installTransform = installTransform;
